Adding OTLP Exporter

codeboten opened the pull request:
This PR adds configuration and retry behaviour specification for the OTLP exporter. Most of the configuration options are based off of the opentelemetry-collector configuration and the environment variables proposed are based on #666 

Fixes #681

carlosalberto commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,21 @@
+
+| Configuration Option | Description                                                  | Default           | Env variable                                                 |
+| -------------------- | ------------------------------------------------------------ | ----------------- | ------------------------------------------------------------ |
+| Endpoint             | target to which the exporter is going to send spans or metrics using the gRPC protocol. The valid syntax is described at https://github.com/grpc/grpc/blob/master/doc/naming.md. | `localhost:55680` | `OTEL_EXPORTER_OTLP_SPAN_ENDPOINT` `OTEL_EXPORTER_OTLP_METRIC_ENDPOINT` |
```
Minor nit: Start the description with uppercase (this is the style used in the rest of the specification tables).

codeboten replied:
Fixed

carlosalberto commented:
This a good start!

codeboten commented:
> Since OTLP supports 3 protocols, grpc, http/json and http/protobufs, it'd be good if there was a configuration option to specify which protocol to use. It can probably just be `protocol` instead of splitting it out into `transport` and `content_type`?

Good point @tsloughter, added an option for protocol.

cijothomas commented on specification/context/api-propagators.md:
```diff
@@ -24,41 +29,94 @@ Table of Contents
-The Propagators API currently consists of one `Format`:
+## Propagator Types
+
+A `Propagator` type defines the restrictions imposed by a specific transport
```
Is this related somehow to the OTLP Exporter?

codeboten replied:
Accientally pushed an update to the branch in my fork, reverted.

cijothomas commented on .github/ISSUE_TEMPLATE/context.md:
```diff
@@ -0,0 +1,13 @@
+---
+name: Context
```
Suggestion: Please use separate PRs for unrelated changes. it makes reviewing a lot easier.

codeboten replied:
Accientally pushed an update to the branch in my fork, reverted.

cijothomas commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,22 @@
+
+## Retry
+
+Transient errors MUST be handled with a retry strategy. This retry strategy MUST implement an exponential back-off with jitter to avoid overwhelming the destination until the network is restored or the destination has recovered.
```
Is there a more detailed description of what is considered Transient error?

codeboten replied:
For gRPC the definition of a transient failure is available here: https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md

I couldn't find an equivalent for HTTP, but I would suggest that any 5xx and at least 408 be treated as transient errors.

cijothomas replied:
If the intention of this spec is to enable all languages to implement the OTLP exporter consistently - then I'd suggest to put the description of 'transient' in this document itself. Else every language can implement it inconsistently. Specifically, which error codes are considered transient for either of the protocols.

codeboten replied:
Added a section for the definition of transient errors

tigrannajaryan replied:
Here is the equivalent for OTLP/HTTP: https://github.com/open-telemetry/oteps/blob/master/text/0099-otlp-http.md#failures

cijothomas commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,21 @@
+# OpenTelemetry Protocol Collector Exporter
+
+This document specifies the configuration options available to the OpenTelemetry Protocol (OTLP) Collector `SpanExporter` and `MetricsExporter` as well as the retry behavior.
```
link to the OTel Collector repo/doc.

codeboten replied:
Updated

cijothomas commented:
> > Since OTLP supports 3 protocols, grpc, http/json and http/protobufs, it'd be good if there was a configuration option to specify which protocol to use. It can probably just be `protocol` instead of splitting it out into `transport` and `content_type`?
> 
> Good point @tsloughter, added an option for protocol.

Seems like this change got lost.

codeboten commented:
> Seems like this change got lost.

It's back now.

cijothomas commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,22 @@
+
+| Configuration Option | Description                                                  | Default           | Env variable                                                 |
+| -------------------- | ------------------------------------------------------------ | ----------------- | ------------------------------------------------------------ |
+| Endpoint             | Target to which the exporter is going to send spans or metrics using the gRPC protocol. The valid syntax is described at https://github.com/grpc/grpc/blob/master/doc/naming.md. | `localhost:55680` | `OTEL_EXPORTER_OTLP_SPAN_ENDPOINT` `OTEL_EXPORTER_OTLP_METRIC_ENDPOINT` |
```
"using the gRPC protocol" - can be omitted as protocol could be different.

mwear replied:
For OTLP/HTTP would the endpoint contain the scheme (http or https)? What should it default to?

codeboten replied:
I would expect the scheme to depend on the value of Insecure for both grpc and http @mwear

tigrannajaryan replied:
For OTLP/HTTP it would be also useful to make the entire URL configurable and not just the host:port portion. This is also described in the spec as a "default" vs "MAY be configured": https://github.com/open-telemetry/oteps/blob/master/text/0099-otlp-http.md#request

codeboten replied:
Updated the description to include an example with a path

nadiaciobanu commented:
Hello! I'm working on the OTLP exporter for OpenTelemetry C++. I was wondering if there's any ETA on the new specification? I'd like to beginning updating the C++ OTLP exporter according to new specification soon.

codeboten commented:
I believe I've addressed all the comments @nadiaciobanu. It would be great if the @open-telemetry/specs-approvers or @open-telemetry/specs-trace-approvers can take another look.

tigrannajaryan commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,41 @@
+| Protocol             | The protocol used to send data to the collector. One of `grpc`,`http/json`,`http/proto`. | `grpc`               | `OTEL_EXPORTER_OTLP_SPAN_PROTOCOL` `OTEL_EXPORTER_OTLP_METRIC_PROTOCOL` |
+| Insecure             | Whether to enable client transport security for the exporter's `grpc` or `http` connection. See [grpc.WithInsecure()](https://godoc.org/google.golang.org/grpc#WithInsecure). | `false`           | `OTEL_EXPORTER_OTLP_SPAN_INSECURE` `OTEL_EXPORTER_OTLP_METRIC_INSECURE` |
+| Certificate File     | Certificate file for TLS credentials of gRPC client. Should only be used if `insecure` is set to `false`. | n/a               | `OTEL_EXPORTER_OTLP_SPAN_CERTIFICATE` `OTEL_EXPORTER_OTLP_METRIC_CERTIFICATE` |
+| Headers              | The headers associated with gRPC requests.                   | n/a               | `OTEL_EXPORTER_OTLP_SPAN_HEADERS` `OTEL_EXPORTER_OTLP_METRIC_HEADERS` |
```
```suggestion
| Headers              | The headers associated with gRPC or HTTP requests.                   | n/a               | `OTEL_EXPORTER_OTLP_SPAN_HEADERS` `OTEL_EXPORTER_OTLP_METRIC_HEADERS` |
```

tigrannajaryan commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,41 @@
+| Insecure             | Whether to enable client transport security for the exporter's `grpc` or `http` connection. See [grpc.WithInsecure()](https://godoc.org/google.golang.org/grpc#WithInsecure). | `false`           | `OTEL_EXPORTER_OTLP_SPAN_INSECURE` `OTEL_EXPORTER_OTLP_METRIC_INSECURE` |
+| Certificate File     | Certificate file for TLS credentials of gRPC client. Should only be used if `insecure` is set to `false`. | n/a               | `OTEL_EXPORTER_OTLP_SPAN_CERTIFICATE` `OTEL_EXPORTER_OTLP_METRIC_CERTIFICATE` |
+| Headers              | The headers associated with gRPC requests.                   | n/a               | `OTEL_EXPORTER_OTLP_SPAN_HEADERS` `OTEL_EXPORTER_OTLP_METRIC_HEADERS` |
+| Compression          | Compression key for supported compression types within collector. | gzip              | `OTEL_EXPORTER_OTLP_SPAN_COMPRESSION` `OTEL_EXPORTER_OTLP_METRIC_COMPRESSION` |
```
What setting should be used to disable the compression?

codeboten replied:
That's a good question. It looks like the otlp exporter in the collector disables compression if its value is `""`, maybe that can be the default? What are your thoughts @tigrannajaryan

tigrannajaryan replied:
How do you pass an empty string in an env variable? I think in many (most) languages empty string is not distinguishable from unspecified value when reading from env. Which mean "gzip" cannot be the default and no compression has to be the default (I assume default means it is the value that is used if you do not specify the env variable).

codeboten replied:
I think we're saying the same thing, the default should be no compression. I've updated the PR.

tigrannajaryan commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,41 @@
+
+## Retry
+
+[Transient errors](#transient-errors) MUST be handled with a retry strategy. This retry strategy MUST implement an exponential back-off with jitter to avoid overwhelming the destination until the network is restored or the destination has recovered.
```
It would be useful to add links to relevant sections of OTLP spec: 
https://github.com/open-telemetry/oteps/blob/master/text/0035-opentelemetry-protocol.md#export-response
https://github.com/open-telemetry/oteps/blob/master/text/0099-otlp-http.md#failures

codeboten replied:
I agree, I didn't see these anywhere in the OTLP spec but it's possible I missed them. Are they anywhere else than in the oteps?

tigrannajaryan replied:
I don't think we have these anywhere else.

codeboten replied:
links have been added

tigrannajaryan commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,41 @@
+# OpenTelemetry Protocol Collector Exporter
+
+This document specifies the configuration options available to the OpenTelemetry Protocol ([OTLP](https://github.com/open-telemetry/opentelemetry-proto)) [Collector](https://github.com/open-telemetry/opentelemetry-collector) `SpanExporter` and `MetricsExporter` as well as the retry behavior.
```
This is probably a better link for OTLP: https://github.com/open-telemetry/oteps/blob/master/text/0035-opentelemetry-protocol.md since it contains the detailed spec and not just the proto definitions.

codeboten replied:
Done

tigrannajaryan commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,41 @@
+
+## Configuration Options
+
+The configuration options are configurable separately for the `SpanExporter` and the `MetricsExporter`.
```
Do we think that spans and metrics are likely to require different endpoints and thus need different env variables? I am guessing the most likely use case will require them to be the same and thus a having separate env vars makes configuration more cumbersome.

codeboten replied:
I think we need the separation at least for http to be able to configure different paths as required for spans & metrics.

tigrannajaryan replied:
OK, makes sense.

tigrannajaryan commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,43 @@
+
+| Configuration Option | Description                                                  | Default           | Env variable                                                 |
+| -------------------- | ------------------------------------------------------------ | ----------------- | ------------------------------------------------------------ |
+| Endpoint             | Target to which the exporter is going to send spans or metrics. This MAY be configured to include a path (e.g. `example.com/v1/traces`). | `localhost:55680` | `OTEL_EXPORTER_OTLP_SPAN_ENDPOINT` `OTEL_EXPORTER_OTLP_METRIC_ENDPOINT` |
```
It is not completely clear what happens if the Endpoint does not include the path. Is the default path automatically appended? So if I specify `Endpoint=example.com` does it mean the actual endpoint is `example.con/v1/trace`? What if I want the path to be the root and don't want the default path to be appended?

codeboten replied:
My expectation here is that the value configured is taken verbatim and passed to the OTLP exporter in the different languages. I would not expect a path to be appended. This does raise a good question of whether a default value for http and another for grpc makes sense here.

tigrannajaryan replied:
I am not sure I like this approach. We are basically not providing any easy way to use the default port and path (for http transport).

The default path for OTLP/HTTP is defined in the spec: https://github.com/open-telemetry/oteps/blob/master/text/0099-otlp-http.md#request 
I think we need to honor it. It is also important to honor the default port if the user does not specify it.

I believe we need to design the env variables such that configuration is easy in the most common case and for the uncommon cases it is OK if it is a bit more difficult.

The most common case is when I need to specify only the ip address or the hostname of the server and choose the transport. Everything else I expect to be set to the correct defaults. The defaults include the port number and when using otlp/http also the path.

For example I would expect the following to result in the destination of `https://example.com:55680/v1/traces`:
```
OTEL_EXPORTER_OTLP_SPAN_ENDPOINT=example.com
OTEL_EXPORTER_OTLP_SPAN_TRANSPORT=http
```

If I wanted to override the path I would probably be OK if I need to specify the full URL using a different env variable and in that case I don't need to specify the ENDPOINT and TRANSPORT:
```
OTEL_EXPORTER_OTLP_SPAN_ENDPOINT_URL=http://example.com:1234/mytraces
```

There may be other ways to structure this but ultimately I want to make sure the most common case is easy and with the current PR I think we are not achieving that goal.

Perhaps it is not possible to achieve what I ask for, so feel free to ignore this comment.

tigrannajaryan commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,43 @@
+# OpenTelemetry Protocol Collector Exporter
```
Is this a specification for a Collector or for any code that implement OTLP protocol client portion? Why do we refer specifically to the Collector here?

codeboten replied:
This is specifically around the OTLP exporter, which the go/js/python implementations refer to as the collector exporter.

tigrannajaryan replied:
I think it is a bit misleading since OTLP is not used by the Collector only and Collector does not use OTLP only. It just happens to be that Collector can also accept OTLP. I think it is better to delete the references to the "Collector" here and in OTLP implementations in the SDKs.

codeboten replied:
Fair enough. Removed the references to the collector from this PR.

tigrannajaryan commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,43 @@
+# OpenTelemetry Protocol Collector Exporter
+
+This document specifies the configuration options available to the OpenTelemetry Protocol ([OTLP](https://github.com/open-telemetry/oteps/blob/master/text/0035-opentelemetry-protocol.md)) [Collector](https://github.com/open-telemetry/opentelemetry-collector) `SpanExporter` and `MetricsExporter` as well as the retry behavior.
```
I am not sure what does `SpanExporter` and `MetricsExporter` refer to in the context of Collector. Just to clarify, there are currently no such concepts in the Collector. Is this a proposal to add these concepts to the Collector?

fbogsany replied:
I think this is an artifact of some SDK implementations providing these separately, and because the processing & export pipelines for metrics and spans in the SDKs are configured separately.

The Go SDK provides a [single OTLP exporter](https://github.com/open-telemetry/opentelemetry-go/tree/master/exporters/otlp) that implements both the `trace.SpanBatcher` and `metric.Exporter` interfaces, and it would be ideal IMO to be able to configure a single instance/connection.

The [Java implementation](https://github.com/open-telemetry/opentelemetry-java/tree/master/exporters/otlp/src/main/java/io/opentelemetry/exporters/otlp) implements the exporters separately and I *think* they use separate connections.

Most other SDKs have only implemented trace export for OTLP so far.

Given that the protocol is designed to multiplex different types of telemetry, I think it'd be good if the spec encouraged implementations to do so over a single connection. This would also simplify configuration, since we wouldn't need to duplicate all the environment variables for `*_SPAN_*` and `*_METRIC_*`.

codeboten replied:
While I agree that it would greatly simplify configuration, we would also be losing the flexibility of sending different types of telemetry data (traces/logs/metrics) to different backends by combining the configuration into a single one. 

Though I suppose a user would still be able to accomplish this in code (by instantiating separate exporters for instance), they just wouldn't be able to do this via environment variable configuration.

fbogsany commented on specification/trace/sdk_exporters/otlp.md:
```diff
@@ -0,0 +1,43 @@
+| Certificate File     | Certificate file for TLS credentials of gRPC client. Should only be used if `insecure` is set to `false`. | n/a               | `OTEL_EXPORTER_OTLP_SPAN_CERTIFICATE` `OTEL_EXPORTER_OTLP_METRIC_CERTIFICATE` |
+| Headers              | The headers associated with gRPC or HTTP requests.                   | n/a               | `OTEL_EXPORTER_OTLP_SPAN_HEADERS` `OTEL_EXPORTER_OTLP_METRIC_HEADERS` |
+| Compression          | Compression key for supported compression types. Supported compression: `gzip`| no compression              | `OTEL_EXPORTER_OTLP_SPAN_COMPRESSION` `OTEL_EXPORTER_OTLP_METRIC_COMPRESSION` |
+| Timeout              | Max waiting time for the backend to process each spans or metrics batch. | 60s               | `OTEL_EXPORTER_OTLP_SPAN_TIMEOUT` `OTEL_EXPORTER_OTLP_METRIC_TIMEOUT` |
```
Some other configuration options are exposed by some exporter implementations. I'm not sure if these are generalizable to implementation in all SDKs, but they're worth considering (if they were previously considered and rejected in this PR, please ignore this comment):

* Go exposes [worker count](https://github.com/open-telemetry/opentelemetry-go/blob/fc1ce6cb8fa42408dcb35f5126fab1f5e4cfe51b/exporters/otlp/options.go#L78-L79) and [reconnection period (after a connection failure)](https://github.com/open-telemetry/opentelemetry-go/blob/fc1ce6cb8fa42408dcb35f5126fab1f5e4cfe51b/exporters/otlp/options.go#L106-L108).
* Ruby exposes [keep alive timeout, open timeout, read timeout, and retry limit](https://github.com/open-telemetry/opentelemetry-ruby/blob/875abed608bd33ef8ba52d765d643c8d2c12c2c9/exporters/otlp/lib/opentelemetry/exporters/otlp/exporter.rb#L36-L39).

Go also allows [configuration of gRPC](https://github.com/open-telemetry/opentelemetry-go/blob/fc1ce6cb8fa42408dcb35f5126fab1f5e4cfe51b/exporters/otlp/options.go#L38-L61), which includes the retry limit (amongst other things).

fbogsany commented:
This PR should also provide clarity about default ports for different protocols. None of the OTEPs for OTLP say anything concrete about default ports. Different SDKs have made incompatible choices.

* [Java](https://github.com/open-telemetry/opentelemetry-java/blob/8d0a4787ff95e7c7861b30dbbf0b42ece133e7cd/exporters/otlp/src/main/java/io/opentelemetry/exporters/otlp/OtlpGrpcSpanExporter.java#L71) uses `localhost:55680` as the default endpoint
* [Python](https://github.com/open-telemetry/opentelemetry-python/blob/3143a4ba16af096584bf1bffbe2d673a87fecc34/exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/trace_exporter/__init__.py#L92) uses `localhost:55678` as the default endpoint (which is actually the default for OpenCensus)
* [Node](https://github.com/open-telemetry/opentelemetry-js/blob/b247e69034a7888a842fe75e0a5ff06f8bea44a8/packages/opentelemetry-exporter-collector/src/platform/node/CollectorTraceExporter.ts#L26-L28) uses `localhost:55680` for gRPC and `http://localhost:55681/v1/trace` for both JSON and JSON/Proto over HTTP
* [C++](https://github.com/open-telemetry/opentelemetry-cpp/blob/d9035ef360bab4a9ed484398d40d463547d9e2e5/exporters/otlp/src/otlp_exporter.cc#L13) uses `localhost:55678` as the default endpoint (which is actually the default for OpenCensus)
* [.NET](https://github.com/open-telemetry/opentelemetry-dotnet/blob/bde64f2b6a78102252270f8d81db06fa33a8fc25/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpExporterOptions.cs#L30) uses `localhost:55680` as the default endpoint
* [Go](https://github.com/open-telemetry/opentelemetry-go/blob/fc1ce6cb8fa42408dcb35f5126fab1f5e4cfe51b/exporters/otlp/options.go#L25-L26) uses 55680 as the default port and `localhost` as the default host
* [The collector receiver](https://github.com/open-telemetry/opentelemetry-collector/tree/master/receiver/otlpreceiver) uses [55680 as the default port for gRPC and 55681 as the default for HTTP](https://github.com/open-telemetry/opentelemetry-collector/blob/bd921afde8e33d46c6f1586b0182cee1addc391e/receiver/otlpreceiver/factory.go#L59-L71)
* Ruby doesn't yet provide a default, mostly because of the lack of clarity in the spec - https://github.com/open-telemetry/opentelemetry-ruby/pull/277

Python and C++ seem to be the outliers here.

There is also contention in the collector about [whether the ports for gRPC and HTTP receivers should be split or unified](https://github.com/open-telemetry/opentelemetry-collector/issues/1256).

codeboten commented:
> 
>     * [Python](https://github.com/open-telemetry/opentelemetry-python/blob/3143a4ba16af096584bf1bffbe2d673a87fecc34/exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/trace_exporter/__init__.py#L92) uses `localhost:55678` as the default endpoint (which is actually the default for OpenCensus)
> 

Good catch for the discrepancy in ports, I suspect it's an artifact of the opencensus receiver having been implemented first, I've created an issue to update this here: https://github.com/open-telemetry/opentelemetry-python/issues/968

fbogsany commented:
It would also be valuable to specify metrics here. When monitoring a tracing pipeline in production, it is useful to have consistent metrics across different SDKs. We have https://github.com/open-telemetry/opentelemetry-ruby/issues/309 open in the Ruby SDK, which suggests:

> At least the following metrics are useful.
> 
> In exporters:
> 
> - `export.failure` - counts failed export requests
> - `export.success` - counts successful export requests
> - `export.spans` - counts spans successfully exported
> 
> In `BatchSpanProcessor` and `SimpleSpanProcessor`:
> 
> - `dropped_spans` - counts spans permanently dropped
>

carlosalberto commented:
> It would also be valuable to specify metrics here.

As this PR and the discussion has grown rather large, I suggest working on them on follow up issues/PRs to merge this ASAP. So let's try to focus on having the minimum, correct set of options that an OTLP exporter MUST define, and iterate over the nice-to-have things (and also more exotic features that may be supported depending on the language).

fbogsany commented:
I was being polite. Sorry. Consistent metrics are a must have for production deployment of OTLP exporters, not a nice to have.

carlosalberto commented:
> Consistent metrics are a must have for production deployment of OTLP exporters, not a nice to have.

Oh, good to know. Still, I'd still suggest to define them in a follow up (we did the same for the old OTEP 66, which changed a lot of context logic, and we had the change done in a few PRs, as it was easier to review and merge).

codeboten commented:
I agree that consistency metrics should be reported. I would expect the metrics to be defined across all exporters and not specifically to the OTLP exporter. Any thoughts around creating a separate issue to track this @fbogsany? Sounds like it could be 2 separate issues:
1. Exporters must report failure/success/spans metrics
2. Processor must report dropped_spans

andrewhsu commented:
@codeboten from the bug triage meeting today, when the ci is green, looks like this can be merged.
@fbogsany can you make sure sure there are follow-up issues to address the items you raised earlier?

codeboten commented:
@andrewhsu sounds good. i'll sort out the git mess i just got myself into and get the CI passing.

carlosalberto commented:
@bogdandrutu Please review this, as it looks ready to be merged.

fbogsany commented:
I'm concerned about the implicit expectation that Span and Metric exporters are configured separately. OTLP is designed to support multiplexing different forms of telemetry (metrics, traces, logs) over a single connection. The Go implementation, at least, encourages configuring it in that way. Doubling (tripling, when we add logs) the configuration variables by requiring them to be specified separately introduces unnecessary complexity into both exporter code (if we continue to support multiplexing, which we should) and service configuration.

The default deployment we should expect, when OTLP is used for both metrics and spans, is to send them both to the same destination (the OTel collector). If someone chooses to use, say, Prometheus for metrics and OTLP for spans, then the configuration of OTLP export should not change - all that changes is that it is not hooked into the metric export pipeline. I would expect OTLP export of metrics and spans to *different* destinations to be the exception rather the common case.

tigrannajaryan commented:
> I'm concerned about the implicit expectation that Span and Metric exporters are configured separately. OTLP is designed to support multiplexing different forms of telemetry (metrics, traces, logs) over a single connection. The Go implementation, at least, encourages configuring it in that way. Doubling (tripling, when we add logs) the configuration variables by requiring them to be specified separately introduces unnecessary complexity into both exporter code (if we continue to support multiplexing, which we should) and service configuration.
> 
> The default deployment we should expect, when OTLP is used for both metrics and spans, is to send them both to the same destination (the OTel collector). If someone chooses to use, say, Prometheus for metrics and OTLP for spans, then the configuration of OTLP export should not change - all that changes is that it is not hooked into the metric export pipeline. I would expect OTLP export of metrics and spans to _different_ destinations to be the exception rather the common case.

I agree with @fbogsany and I had the [same concern](https://github.com/open-telemetry/opentelemetry-specification/pull/699#discussion_r459173485).

From the number of upvotes I see this is a shared concern.

At the risk of repeating what Francis said: my expectation would be that it is should be possible to configure a single ENDPOINT env variable to which then all telemetry is sent (traces, metrics and in the future logs as well). This is very likely going to be the most common use case, with the ENDPOINT likely pointing to the Collector. As a consequence all other options, such as compression, transport, certificate, etc are also likely going to be the same for all signal types.

Requireng that each of the signals repeats all this env variables makes the most common use case complicated to use. We should aim to make this most frequent case easy.

I understand that sometimes (likely rarely) we may want to send different signals to different endpoints or use different options. This is rare and it is ok if it is more cumbersome to configure (e.g. with additional env variables per signal which override the common ones).

For the sake of making progress I am happy if this PR gets merged as is provided that we create an issue to address this concern before GA. @codeboten if you are OK with this approach please create the issue.

codeboten commented:
>For the sake of making progress I am happy if this PR gets merged as is provided that we create an issue to address this concern before GA. @codeboten if you are OK with this approach please create the issue.

@tigrannajaryan I've opened follow up issues regarding the OTLP exporter, happy to tackle #790 right away. @fbogsany can you open an issue to address the capturing metrics in the span processors?

carlosalberto commented:
> I understand that sometimes (likely rarely) we may want to send different signals to different endpoints or use different options. This is rare and it is ok if it is more cumbersome to configure (e.g. with additional env variables per signal which override the common ones).

This sounds like a good solution IMHO.

tigrannajaryan commented:
> @tigrannajaryan I've opened follow up issues regarding the OTLP exporter, happy to tackle #790 right away.

@codeboten Thank you. Up to you if you want to do it in this PR or in a follow up PR.

carlosalberto commented:
Hey @fbogsany sounds ok to merge this one and work on the remaining issues in [the follow up](#790)?

@tigrannajaryan Please approve ;)

fbogsany commented:
I 👍 earlier comments, but that may not have been visible enough. This LGTM. I'll start a PR next week for the span processor metrics.
